File: day20/day20.py
with open('input.txt', 'r') as f:
    raw_input = f.read()

encrypted = [int(i) for i in raw_input.splitlines()]

decrypted = encrypted.copy()

length = len(encrypted)

for i, num in enumerate(encrypted):
    old_pos = decrypted.index(num)
    new_pos = (num + old_pos) % (length-1)
    decrypted = decrypted[:old_pos] + decrypted[old_pos+1:]
    if new_pos == 0:
        decrypted.append(num)
    else:
        decrypted.insert(new_pos, num)

zero_pos = decrypted.index(0)
print(decrypted[(zero_pos+1000)%length])
print(decrypted[(zero_pos+2000)%length])
print(decrypted[(zero_pos+3000)%length])
grove_sum = decrypted[(zero_pos+1000)%length] + decrypted[(zero_pos+2000)%length] + decrypted[(zero_pos+3000)%length]
print(grove_sum)

File: day20/day20.hs
import Data.List

decrypt :: [Int] -> [Int]
decrypt = id

-- moves element at index a to index b
move :: Int -> Int -> [Int] -> [Int]
move a b list = insertAt (list !! a) b . removeAt a $ list

enumerate :: [Int] -> [(Int, Int)]
enumerate = zip [0..]

removeAt :: Int -> [Int] -> [Int]
removeAt index  = map snd . filter (\(i, _) -> i /= index) . enumerate

insertAt :: Int -> Int -> [Int] -> [Int]
insertAt el 0 xs = el : xs
insertAt el i (x:xs) = x : insertAt el (i-1) xs

-- elemIndex idx list


main = do
    input <- readFile "test.txt"
    let encrypted = map read $ lines input
    let decrypted = move 0 2 encrypted
    print decrypted